hive/agents/ppo.py

```diff
@@ -0,0 +1,411 @@
+                    ratios = torch.exp(logratios) 
+                    advantages = batch["advantages"]
+                    if self._normalize_advantages:
+                        advantages = (advantages - advantages.mean()) / (advantages.std() + 1e-8)
+                    # Actor loss
+                    loss1 = -advantages * ratios
+                    loss2 = -advantages * torch.clamp(ratios, 1 - self._clip_coef, 1 + self._clip_coef)
+                    loss = torch.max(loss1, loss2).mean()
+                    entr_loss = entropy.mean()
+                    actor_loss = loss - self._ent_coef * entr_loss
+                    actor_loss.backward()
+                    if self._grad_clip is not None:
+                        torch.nn.utils.clip_grad_norm_(
+                            self._actor.parameters(), self._grad_clip
+                        )
+                    self._actor_optimizer.step()
+
+                    # Critic loss
+                    values = values.view(-1)
+                    if self._clip_vloss:
+                        v_loss_unclipped = self._critic_loss_fn(values, batch["returns"])
+                        v_clipped = batch["values"] + torch.clamp(values - batch["values"], 
+                            -self._clip_coef,
+                            self._clip_coef,
+                        )
+                        v_loss_clipped = self._critic_loss_fn(v_clipped, batch["returns"])
+                        v_loss_max = torch.max(v_loss_unclipped, v_loss_clipped)
+                        critic_loss = 0.5 * v_loss_max.mean()
+                    else:
+                        critic_loss = 0.5 * self._critic_loss_fn(values, batch["returns"]).mean()
+                    critic_loss = self._vf_coef * critic_loss
+                    critic_loss.backward()
+                    if self._grad_clip is not None:
+                        torch.nn.utils.clip_grad_norm_(
+                            self._critic.parameters(), self._grad_clip
+                        )
+                    self._critic_optimizer.step()
+
+                    with torch.no_grad():
+                        # calculate approx_kl http://joschu.net/blog/kl-approx.html
+                        approx_kl = ((ratios - 1) - logratios).mean()
+
+                    if self._logger.should_log(self._timescale):
+                        self._logger.log_scalar("actor_loss", actor_loss, self._timescale)
+                        self._logger.log_scalar("critic_loss", critic_loss, self._timescale)
+                        self._logger.log_scalar("entropy_loss", entr_loss, self._timescale)
+                        self._logger.log_scalar("approxkl", approx_kl, self._timescale)
+                
+                if self._target_kl is not None:
+                    if approx_kl > self._target_kl:
+                        break
+            self._replay_buffer.reset()
+
+        processed_update_info = self.preprocess_update_info(update_info)
```

Comment:
This is being done to get the next_obs to bootstrap the value estimte to calculate advantages. Issue with this approach is a transition from an older policy is used for the next update. 

@dapatil211, would it be possible to get the next_obs in the update info from the single agent runner loop?